haskell/test/Test/Ligo/BaseDAO/OffChainViews.hs

```diff
@@ -44,24 +34,24 @@ test_FA2 :: TestTree
 test_FA2 =
   mkFA2Tests offChainViewStorage (mkMetadataSettings defaultMetadataConfig)
 
-runView
-  :: forall ret. (HasCallStack, IsoValue ret)
-  => View $ ToT Storage
-  -> Storage
-  -> ViewParam
-  -> Either MichelsonFailed ret
-runView view storage param =
-  case interpretView dummyContractEnv view param storage of
-    Right x -> Right x
-    Left (VIEMichelson _ (MSVIEMichelsonFailed (MichelsonFailureWithStack e _))) -> Left e
-    Left err -> error (pretty err)
-
+-- runView
+--   :: forall ret mp. (HasCallStack, IsoValue ret)
+--   => View $ ToT Storage
+--   -> Storage
+--   -> ViewParam mp
+--   -> Either MichelsonFailed ret
+-- runView view storage param =
+--   case interpretView dummyContractEnv view param storage of
+--     Right x -> Right x
+--     Left (VIEMichelson _ (MSVIEMichelsonFailed (MichelsonFailureWithStack e _))) -> Left e
+--     Left err -> error (pretty err)
 
 mkFA2Tests :: Storage -> MetadataSettings -> TestTree
-mkFA2Tests storage mc = testGroup "FA2 off-chain views"
-  [ testGroup "governance_token" $
-    [ testCase "Get the address and token_id of the associated FA2 contract" $
-        runView @GovernanceToken (governanceTokenView mc) storage NoParam
-        @?= (Right $ GovernanceToken genesisAddress FA2.theTokenId)
-    ]
-  ]
+mkFA2Tests _ _ = testGroup "FA2 off-chain views" []
+-- TODO: Uncomment when some way to run views is available.
+--   [ testGroup "governance_token" $
+--     [ testCase "Get the address and token_id of the associated FA2 contract" $
+--         runView @GovernanceToken (governanceTokenView mc) storage NoParam
+--         @?= (Right $ GovernanceToken (MkAddress genesisAddress) FA2.theTokenId)
+--     ]
+--   ]
```

Comment:
Correct me if I'm wrong, but there's certainly a way to run this test without `interpretView`? Just convert it to a Cleveland test, originate the contract with this view and use `callOffChainView`?

Reply:
I gave this a try, and found myself having a hard time getting `MetadataMap` from `Metadata` (provided by `knownBaseDAOMetadata`)  to populate the views in `offchainViewStorage`.

Reply:
Uh. So, here's the diff that _should_ work (notice modules are imported wherever):

```diff
diff --git a/haskell/test/Test/Ligo/BaseDAO/OffChainViews.hs b/haskell/test/Test/Ligo/BaseDAO/OffChainViews.hs
index f273899..d0679eb 100644
--- a/haskell/test/Test/Ligo/BaseDAO/OffChainViews.hs
+++ b/haskell/test/Test/Ligo/BaseDAO/OffChainViews.hs
@@ -1,11 +1,14 @@
 -- SPDX-FileCopyrightText: 2021 Tezos Commons
 -- SPDX-License-Identifier: LicenseRef-MIT-TC
 
+{-# LANGUAGE OverloadedLists #-}
+
 module Test.Ligo.BaseDAO.OffChainViews
   ( test_FA2
   ) where
 
 import Universum hiding (view)
+import Lorentz (Contract, defaultContract, cdr, nil, pair, (#), NiceStorage, HasAnnotation)
 
 import Morley.Michelson.Runtime.GState (genesisAddress)
 import Morley.Tezos.Address
@@ -14,13 +17,23 @@ import Test.Tasty (TestTree, testGroup)
 
 import Ligo.BaseDAO.Types
 import Ligo.BaseDAO.TZIP16Metadata
+import Lorentz.Contracts.Spec.FA2Interface qualified as FA2
+import Lorentz.Contracts.Spec.TZIP16Interface
+import Test.Morley.Metadata
+import Test.Cleveland
+import Data.Aeson (encode)
+import Data.ByteString.Lazy qualified as BS
+
+metadataBigMap :: MetadataConfig -> MetadataMap
+metadataBigMap conf = metadataURI (tezosStorageUri selfHost "metadata")
+  <> [("metadata", BS.toStrict $ encode $ knownBaseDAOMetadata $ mkMetadataSettings conf)]
 
 offChainViewStorage :: Storage
 offChainViewStorage =
   (mkStorage
   ! #admin addr
   ! #extra ()
-  ! #metadata mempty
+  ! #metadata (metadataBigMap defaultMetadataConfig)
   ! #level 100
   ! #tokenAddress (MkAddress genesisAddress)
   ! #quorumThreshold (mkQuorumThreshold 1 100)
@@ -31,27 +44,10 @@ offChainViewStorage =
     addr = [ta|tz1M6dcor9QNTFr9Ri68cBYvpxrogZaMttuE|]
 
 test_FA2 :: TestTree
-test_FA2 =
-  mkFA2Tests offChainViewStorage (mkMetadataSettings defaultMetadataConfig)
-
--- runView
---   :: forall ret mp. (HasCallStack, IsoValue ret)
---   => View $ ToT Storage
---   -> Storage
---   -> ViewParam mp
---   -> Either MichelsonFailed ret
--- runView view storage param =
---   case interpretView dummyContractEnv view param storage of
---     Right x -> Right x
---     Left (VIEMichelson _ (MSVIEMichelsonFailed (MichelsonFailureWithStack e _))) -> Left e
---     Left err -> error (pretty err)
-
-mkFA2Tests :: Storage -> MetadataSettings -> TestTree
-mkFA2Tests _ _ = testGroup "FA2 off-chain views" []
--- TODO: Uncomment when some way to run views is available.
---   [ testGroup "governance_token" $
---     [ testCase "Get the address and token_id of the associated FA2 contract" $
---         runView @GovernanceToken (governanceTokenView mc) storage NoParam
---         @?= (Right $ GovernanceToken (MkAddress genesisAddress) FA2.theTokenId)
---     ]
---   ]
+test_FA2 = testScenarioOnEmulator "FA2 off-chain views" $ scenario $ do
+  contract <- originate "test_FA2" offChainViewStorage (idContract @Storage)
+  callOffChainView @GovernanceToken contract "governance_token" NoParam
+    @@== (GovernanceToken (MkAddress genesisAddress) FA2.theTokenId)
+
+idContract :: (NiceStorage st, HasAnnotation st) => Contract () st ()
+idContract = defaultContract $ cdr # nil # pair

```

However, I've spent the last 45 minutes or so staring at GHC struggling to compile baseDAO tests, so I don't know if this works. Was it always this slow? I mean, the last build took a full hour on CI.

Reply:
```
*** Desugar [Test.Ligo.BaseDAO.Management.TransferOwnership]:
Result size of Desugar (before optimization)
  = {terms: 19,984,
     types: 168,417,
     coercions: 2,119,458,
     joins: 0/5,488}
Result size of Desugar (after optimization)
  = {terms: 16,309,
     types: 152,046,
     coercions: 2,137,815,
     joins: 10/3,905}
!!! Desugar [Test.Ligo.BaseDAO.Management.TransferOwnership]: finished in 35340.72 milliseconds, allocated 95255.164 megabytes
```

> coercions: 2,119,458,
> allocated 95255.164 megabytes

WTH

Reply:
My current implimentation is almost the same as your diff, but returns the following error for some reason. Trying your diff now..

```
governanceToken <- callOffChainView @GovernanceToken baseDao "governance_token" NoParam
                     ┃                              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
                     ┃                              | Metadata bigmap does not contain the key: ''
```

Reply:
> Was it always this slow? I mean, the last build took a full hour on CI.

Yes, AFAIU it became this slow when we had to share tests for `registryDAO` between its native version and the `LambdaDAO` version. So these tests were changed to be polymorphic over storage and parameter. If you can think of a better way to solve this (that is reuse the tests between two contracts with slightly different parameter/storage), please let us know. Slow compilation has been such a pain for a while now..

Reply:
Worked. Thanks!

Reply:
So... It's still slow as heck, but _apparently_ GHC really dislikes all those partial matches everywhere. I've got rid of those using `SizedList`, and now it's still slow, but it got from "I'll die sooner than this finishes" to "I hate it, but I can live with it" slow. To put some specific numbers to that, with `--ghc-options=-j`:

```
real    5m31,311s
user    19m36,553s
sys     3m49,918s
```

I'll push the changes I made to some branch here if I have commit rights.

Reply:
So the slowdown was not due to the large number of constraints on the polymorphic tests? That is comforting, but please, when you have a minute, can you share how you figured it?

>I'll push the changes I made to some branch here if I have commit rights.

For the timebeing, can you please send me a patch from `git` that I can import?

Reply:
> can you share how you figured it?

Eh. How I _could_ have figured it out is by looking at the core output. How I _did_... Well, a lucky educated guess, frankly. Since running the compiler with `-v2` shows it gets very stuck on the desugarer phase, and reports an ungodly amount of memory allocations, this suggests it generates _huge_ core output. I vaguely remembered that partial pattern matches are supposed to turn into explicit errors in the desugarer, so I suspected there might be some issues with that. Getting rid of partial pattern matches was the most obvious thing to try at that point, really. TBH, it's mostly guesswork until I tried and seen some improvement.

Apparently I do have commit rights, so I've pushed `lierdakil/try-use-sized-list`

Reply:
Fair warning: the code I pushed is messy, it needs a bit of clean-up.

Reply:
The compile time has dropped from 30 minutes to 2 minutes 40 secs for a recompiling to account for change in a single module. I cannot figure how this slowdown could be caused by partial pattern matches alone. Is it possible that the constraints somehow play a role in this? becuase IIRC introducing them triggered the slow down, when the partial patterns were there from the begining ..

Anyway this change takes away some massive pain, so a big thanks for figuring this out.